File: ios/PhotoBackup/RemoteLibrary.swift
import Foundation
import Photos

struct RemoteResource: Decodable, Identifiable {
    let resourceId: UUID
    let role: String
    let filename: String
    let mimeType: String
    let contentSize: UInt64
    let storageEncoding: String
    let contentPath: String

    var id: UUID { resourceId }
}

struct RemoteAsset: Decodable, Identifiable {
    let assetId: UUID
    let sourceCreatedAtMs: Int64
    let favorite: Bool
    let archived: Bool
    let trashedAtMs: Int64?
    let tagNames: [String]
    let resources: [RemoteResource]

    var id: UUID { assetId }
    var isTrashed: Bool { trashedAtMs != nil }
    var primary: RemoteResource? {
        resources.first(where: { $0.role == "primary" })
            ?? resources.first(where: { $0.role != "thumbnail" })
    }
    var thumbnail: RemoteResource? { resources.first(where: { $0.role == "thumbnail" }) }
}

private struct TimelinePage: Decodable {
    let items: [RemoteAsset]
    let nextCursor: String?
}

private struct SyncPage: Decodable {
    let nextSequence: Int64
    let hasMore: Bool
}

struct RemoteLibrary {
    let serverURL: URL
    let token: String

    private let decoder: JSONDecoder = {
        let value = JSONDecoder()
        value.keyDecodingStrategy = .convertFromSnakeCase
        return value
    }()

    func timeline(trashed: Bool) async throws -> [RemoteAsset] {
        var result: [RemoteAsset] = []
        var cursor: String?
        repeat {
            var components = URLComponents(url: serverURL.appending(path: "/v1/timeline"), resolvingAgainstBaseURL: false)
            components?.queryItems = [
                URLQueryItem(name: "trashed", value: trashed ? "true" : "false"),
                URLQueryItem(name: "limit", value: "100"),
                cursor.map { URLQueryItem(name: "cursor", value: $0) },
            ].compactMap { $0 }
            guard let url = components?.url else { throw RemoteLibraryError.invalidURL }
            let (data, response) = try await URLSession.shared.data(for: authorized(url: url, method: "GET"))
            try requireSuccess(response, data: data)
            let page = try decoder.decode(TimelinePage.self, from: data)
            result.append(contentsOf: page.items)
            guard page.nextCursor == nil || page.nextCursor != cursor else {
                throw RemoteLibraryError.invalidCursor
            }
            cursor = page.nextCursor
        } while cursor != nil
        return result
    }

    func advanceSync(from initialSequence: Int64) async throws -> Int64 {
        var sequence = initialSequence
        var more: Bool
        repeat {
            var components = URLComponents(url: serverURL.appending(path: "/v1/sync"), resolvingAgainstBaseURL: false)
            components?.queryItems = [
                URLQueryItem(name: "after", value: String(sequence)),
                URLQueryItem(name: "limit", value: "1000"),
            ]
            guard let url = components?.url else { throw RemoteLibraryError.invalidURL }
            let (data, response) = try await URLSession.shared.data(for: authorized(url: url, method: "GET"))
            try requireSuccess(response, data: data)
            let page = try decoder.decode(SyncPage.self, from: data)
            sequence = page.nextSequence
            more = page.hasMore
        } while more
        return sequence
    }

    func setFavorite(asset: RemoteAsset, value: Bool) async throws {
        try await sendJSON(path: "/v1/assets/\(asset.assetId)", method: "PATCH", body: ["favorite": value])
    }

    func setArchived(asset: RemoteAsset, value: Bool) async throws {
        try await sendJSON(path: "/v1/assets/\(asset.assetId)", method: "PATCH", body: ["archived": value])
    }

    func addTag(named rawName: String, to asset: RemoteAsset) async throws {
        let name = rawName.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !name.isEmpty else { return }
        let (listData, listResponse) = try await URLSession.shared.data(for: authorized(
            url: serverURL.appending(path: "/v1/tags"),
            method: "GET"
        ))
        try requireSuccess(listResponse, data: listData)
        let tags = try decoder.decode([RemoteTag].self, from: listData)
        let tag: RemoteTag
        if let existing = tags.first(where: { $0.name.caseInsensitiveCompare(name) == .orderedSame }) {
            tag = existing
        } else {
            var request = authorized(url: serverURL.appending(path: "/v1/tags"), method: "POST")
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: ["name": name])
            let (data, response) = try await URLSession.shared.data(for: request)
            try requireSuccess(response, data: data)
            tag = try decoder.decode(RemoteTag.self, from: data)
        }
        try await send(path: "/v1/tags/\(tag.tagId)/assets/\(asset.assetId)", method: "POST")
    }

    func duplicateGroupCount() async throws -> Int {
        let (data, response) = try await URLSession.shared.data(for: authorized(
            url: serverURL.appending(path: "/v1/duplicates"),
            method: "GET"
        ))
        try requireSuccess(response, data: data)
        return (try JSONSerialization.jsonObject(with: data) as? [[String: Any]] ?? []).count
    }

    func thumbnailData(for asset: RemoteAsset) async throws -> Data? {
        guard let thumbnail = asset.thumbnail, thumbnail.storageEncoding == "plain-v1" else { return nil }
        let (data, response) = try await URLSession.shared.data(for: authorized(
            url: serverURL.appending(path: thumbnail.contentPath),
            method: "GET"
        ))
        try requireSuccess(response, data: data)
        return data
    }

    func trash(asset: RemoteAsset) async throws {
        try await send(path: "/v1/assets/\(asset.assetId)/trash", method: "POST")
    }

    func restoreFromTrash(asset: RemoteAsset) async throws {
        try await send(path: "/v1/assets/\(asset.assetId)/restore", method: "POST")
    }

    func restoreToPhotos(asset: RemoteAsset) async throws -> String {
        guard let resource = asset.primary else { throw RemoteLibraryError.noPrimaryResource }
        guard resource.storageEncoding == "plain-v1" else { throw RemoteLibraryError.legacyEncryptedResource }
        let suffix = URL(fileURLWithPath: resource.filename).pathExtension
        let temporary = FileManager.default.temporaryDirectory
            .appendingPathComponent(UUID().uuidString)
            .appendingPathExtension(suffix.isEmpty ? "bin" : suffix)
        defer { try? FileManager.default.removeItem(at: temporary) }

        let (downloaded, response) = try await URLSession.shared.download(for: authorized(
            url: serverURL.appending(path: resource.contentPath),
            method: "GET"
        ))
        try requireSuccess(response, data: Data())
        try FileManager.default.moveItem(at: downloaded, to: temporary)
        try await PHPhotoLibrary.shared().performChanges {
            if resource.mimeType.hasPrefix("video/") {
                PHAssetChangeRequest.creationRequestForAssetFromVideo(atFileURL: temporary)
            } else {
                PHAssetChangeRequest.creationRequestForAssetFromImage(atFileURL: temporary)
            }
        }
        return resource.filename
    }

    private func send(path: String, method: String) async throws {
        let (data, response) = try await URLSession.shared.data(for: authorized(
            url: serverURL.appending(path: path),
            method: method
        ))
        try requireSuccess(response, data: data)
    }

    private func sendJSON(path: String, method: String, body: [String: Any]) async throws {
        var request = authorized(url: serverURL.appending(path: path), method: method)
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, response) = try await URLSession.shared.data(for: request)
        try requireSuccess(response, data: data)
    }

    private func authorized(url: URL, method: String) -> URLRequest {
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        return request
    }

    private func requireSuccess(_ response: URLResponse, data: Data) throws {
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw RemoteLibraryError.server(String(decoding: data, as: UTF8.self))
        }
    }
}

private struct RemoteTag: Decodable {
    let tagId: UUID
    let name: String
}

enum RemoteLibraryError: LocalizedError {
    case invalidURL
    case noPrimaryResource
    case legacyEncryptedResource
    case invalidCursor
    case server(String)

    var errorDescription: String? {
        switch self {
        case .invalidURL: "服务器地址无效"
        case .noPrimaryResource: "该资产没有可恢复的原始资源"
        case .legacyEncryptedResource: "旧版加密资源需要旧客户端解密后重新上传"
        case .invalidCursor: "服务器返回了无效的分页游标"
        case .server(let message): message.isEmpty ? "服务器请求失败" : message
        }
    }
}

File: clients/android/app/src/main/java/com/example/mediabackup/RemoteLibrary.kt
package com.example.mediabackup.v02

import android.content.ContentValues
import android.content.Context
import android.os.Build
import android.provider.MediaStore
import org.json.JSONObject

data class RemoteResource(
    val id: String,
    val role: String,
    val filename: String,
    val mimeType: String,
    val contentPath: String,
)

data class RemoteAsset(
    val id: String,
    val createdAtMs: Long,
    val favorite: Boolean,
    val archived: Boolean,
    val trashed: Boolean,
    val tagNames: List<String>,
    val resources: List<RemoteResource>,
)

object RemoteLibrary {
    fun advanceSync(api: BackupApi, initialSequence: Long): Long {
        var sequence = initialSequence
        var more: Boolean
        do {
            val page = api.sync(sequence)
            sequence = page.getLong("next_sequence")
            more = page.getBoolean("has_more")
        } while (more)
        return sequence
    }

    fun timeline(api: BackupApi, trashed: Boolean = false): List<RemoteAsset> {
        val result = mutableListOf<RemoteAsset>()
        var cursor: String? = null
        do {
            val page = api.timeline(cursor = cursor, trashed = trashed)
            val items = page.getJSONArray("items")
            for (position in 0 until items.length()) result += parseAsset(items.getJSONObject(position))
            val next = page.optString("next_cursor").takeIf { it.isNotBlank() && it != "null" }
            check(next == null || next != cursor) { "服务器返回了无效的分页游标" }
            cursor = next
        } while (cursor != null)
        return result
    }

    fun restore(context: Context, api: BackupApi, asset: RemoteAsset): String {
        val resource = asset.resources.firstOrNull { it.role == "primary" }
            ?: asset.resources.firstOrNull { it.role != "thumbnail" }
            ?: error("该资产没有可恢复的原始资源")
        val collection = if (resource.mimeType.startsWith("video/")) {
            MediaStore.Video.Media.EXTERNAL_CONTENT_URI
        } else {
            MediaStore.Images.Media.EXTERNAL_CONTENT_URI
        }
        val values = ContentValues().apply {
            put(MediaStore.MediaColumns.DISPLAY_NAME, resource.filename)
            put(MediaStore.MediaColumns.MIME_TYPE, resource.mimeType)
            if (Build.VERSION.SDK_INT >= 29) {
                put(MediaStore.MediaColumns.RELATIVE_PATH, "Pictures/MediaBackup Restored")
                put(MediaStore.MediaColumns.IS_PENDING, 1)
            }
        }
        val uri = context.contentResolver.insert(collection, values) ?: error("无法创建恢复文件")
        try {
            context.contentResolver.openOutputStream(uri)?.use { api.download(resource.contentPath, it) }
                ?: error("无法打开恢复文件")
            if (Build.VERSION.SDK_INT >= 29) {
                context.contentResolver.update(uri, ContentValues().apply { put(MediaStore.MediaColumns.IS_PENDING, 0) }, null, null)
            }
            return resource.filename
        } catch (error: Exception) {
            context.contentResolver.delete(uri, null, null)
            throw error
        }
    }

    fun thumbnail(api: BackupApi, asset: RemoteAsset): ByteArray? {
        val resource = asset.resources.firstOrNull { it.role == "thumbnail" }
            ?: return null
        return api.downloadBytes(resource.contentPath)
    }

    private fun parseAsset(value: JSONObject): RemoteAsset {
        val resources = value.getJSONArray("resources")
        return RemoteAsset(
            id = value.getString("asset_id"),
            createdAtMs = value.getLong("source_created_at_ms"),
            favorite = value.getBoolean("favorite"),
            archived = value.getBoolean("archived"),
            trashed = !value.isNull("trashed_at_ms"),
            tagNames = buildList {
                val tags = value.optJSONArray("tag_names") ?: org.json.JSONArray()
                for (position in 0 until tags.length()) add(tags.getString(position))
            },
            resources = buildList {
                for (position in 0 until resources.length()) {
                    val resource = resources.getJSONObject(position)
                    check(resource.getString("storage_encoding") == "plain-v1") {
                        "服务器返回了非当前存储协议"
                    }
                    add(
                        RemoteResource(
                            id = resource.getString("resource_id"),
                            role = resource.getString("role"),
                            filename = resource.getString("filename"),
                            mimeType = resource.getString("mime_type"),
                            contentPath = resource.getString("content_path"),
                        ),
                    )
                }
            },
        )
    }
}
